3_summarize/src/annual_thermal_metrics.R

```diff
@@ -0,0 +1,256 @@
+
+calculate_annual_metrics <- function(target_name, site_ids, file_pattern) {
+  
+  purrr::map(site_ids, function(id) {
+    
+    data_ready <- read_feather(sprintf(file_pattern, id)) %>% 
+      select(site_id, date = DateTime, starts_with("temp_")) %>% 
+      pivot_longer(cols = starts_with("temp_"), names_to = "depth", values_to = "wtr") %>% 
+      mutate(depth = as.numeric(gsub("temp_", "", depth))) %>% 
+      mutate(year = as.numeric(format(date, "%Y")))
+    
+    annual_metrics <- data_ready %>% 
+      group_by(site_id, year) %>% 
+      summarize(
+        # winter_dur_0_4 = winter_dur_0_4(),
+        # coef_var_30_60 = coef_var_30_60(),
+        # coef_var_0_30 = coef_var_0_30(),
+        # stratification_onset_yday = stratification_onset_yday(),
+        # stratification_duration = stratification_duration(),
+        # sthermo_depth_mean = sthermo_depth_mean(),
+        
+        peak_temp = peak_temp(date, wtr, depth),
+        gdd_wtr_0c = gdd_wtr_0c(date, wtr),
+        gdd_wtr_5c = gdd_wtr_5c(date, wtr),
+        gdd_wtr_10c = gdd_wtr_10c(date, wtr),
+        
+        # bottom_temp_at_strat = bottom_temp_at_strat(),
+        # schmidt_daily_annual_sum = schmidt_daily_annual_sum(),
+        
+        mean_surf_jas = mean_surf_jas(date, wtr, depth),
+        max_surf_jas = max_surf_jas(date, wtr, depth),
+        mean_bot_jas = mean_bot_jas(date, wtr, depth),
+        max_bot_jas = max_bot_jas(date, wtr, depth),
+        
+        # Per month calcs return a data.frame and get "unpacked" below 
+        mean_surf_mon = mean_surf_mon(date, wtr, depth),
+        max_surf_mon = max_surf_mon(date, wtr, depth),
+        mean_bot_mon = mean_bot_mon(date, wtr, depth),
+        max_bot_mon = max_bot_mon(date, wtr, depth),
+        
+        .groups = "keep" # suppresses message about regrouping
+      ) %>% 
+      unpack(cols = c(mean_surf_mon, max_surf_mon, 
+                      mean_bot_mon, max_bot_mon))
+    
+    message(sprintf("Completed annual metrics for %s/%s", which(site_ids == id), length(site_ids)))
+    
+    return(annual_metrics)
+  }) %>% 
+    bind_rows() %>% 
+    readr::write_csv(target_name)
+  
+}
+
+# Collection of functions to calculate annual thermal metrics
+# All accept the same main data input, `date` and `wtr`.
+
+# It is assumed that these functions only get passed one
+# years worth of data.
+
+#' @param date a vector (or one column) with the dates of class `Date`
+#' @param wtr a vector (or one column) with water temperature values in deg C
+
+# Some also have the following parameters:
+#' @param depth vector (or column) with depth in meters
+
+#' @description # of days, beginning October 30 of the previous year and running 
+#' through June of the given year, that surface water is <4 degrees C
+winter_dur_0_4 <- function(date, wtr) {
+  
+}
+
+#' @description Coefficient of Variation of surface temperature from 30-60 days post ice off.
+coef_var_30_60 <- function(date, wtr) {
```

Comment:
Wondering about the use of making these `coef_var_post_ice()` w/ `days_past` or something like that as inputs. I'm also OK leaving these two as-is, since I think we only use these two post-ice metics for COV

Reply:
I went through and generalized the other functions to make them more flexible. Will plan to do that for these once I get to implementing.